<a href="https://colab.research.google.com/github/RobsonGFerrarezi/Desafio_aula_12-04/blob/main/Exercicio_aula_12_04.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

<a rel="license" href="https://faculdadesalvadorarena.org.br/"><img alt="FESA" style="border-width:0" src="https://faculdadesalvadorarena.org.br/wp-content/uploads/2022/07/logo_fesa.png" /></a><br />
**FESA - Eletiva II (2025.1)** <br/>

Aluno(a): Robson Guilherme Ferrarezi <br/>
Data: 12/04/2025 <br/>

Importando biblioteca Pandas, para manipulação e análise de dados

In [21]:
import pandas as pd

Puxando base de dados, e criando o DataFrame

In [22]:
!wget --no-check-certificate --content-disposition 'https://raw.githubusercontent.com/RobsonGFerrarezi/Desafio_aula_12-04/refs/heads/main/2004-2021.tsv'

--2025-04-12 14:06:26--  https://raw.githubusercontent.com/RobsonGFerrarezi/Desafio_aula_12-04/refs/heads/main/2004-2021.tsv
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.109.133, 185.199.108.133, 185.199.111.133, ...
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.109.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 15640047 (15M) [text/plain]
Saving to: ‘2004-2021.tsv.1’

2004-2021.tsv.1     100%[===================>]  14.92M  --.-KB/s    in 0.1s    

2025-04-12 14:06:26 (124 MB/s) - ‘2004-2021.tsv.1’ saved [15640047/15640047]



In [23]:
df = pd.read_csv('2004-2021.tsv', sep='\t')

In [24]:
df.head()

DATA INICIAL  DATA FINAL        REGIÃO              ESTADO  \
0   2004-05-09  2004-05-15  CENTRO OESTE    DISTRITO FEDERAL   
1   2004-05-09  2004-05-15  CENTRO OESTE               GOIAS   
2   2004-05-09  2004-05-15  CENTRO OESTE         MATO GROSSO   
3   2004-05-09  2004-05-15  CENTRO OESTE  MATO GROSSO DO SUL   
4   2004-05-09  2004-05-15      NORDESTE             ALAGOAS   

            PRODUTO  NÚMERO DE POSTOS PESQUISADOS UNIDADE DE MEDIDA  \
0  ETANOL HIDRATADO                           127              R$/l   
1  ETANOL HIDRATADO                           387              R$/l   
2  ETANOL HIDRATADO                           192              R$/l   
3  ETANOL HIDRATADO                           162              R$/l   
4  ETANOL HIDRATADO                           103              R$/l   

   PREÇO MÉDIO REVENDA  DESVIO PADRÃO REVENDA  PREÇO MÍNIMO REVENDA  \
0                1.288                  0.016                  1.19   
1                1.162                  0.114                  0.89   
2                1.389                  0.097                  1.18   
3                1.262                  0.070                  1.09   
4                1.181                  0.078                  1.05   

   PREÇO MÁXIMO REVENDA MARGEM MÉDIA REVENDA  COEF DE VARIAÇÃO REVENDA  \
0                 1.350                0.463                     0.012   
1                 1.449                0.399                     0.098   
2                 1.760                0.419                     0.070   
3                 1.509                0.432                     0.055   
4                 1.400                 0.24                     0.066   

  PREÇO MÉDIO DISTRIBUIÇÃO DESVIO PADRÃO DISTRIBUIÇÃO  \
0                    0.825                       0.11   
1                    0.763                      0.088   
2                     0.97                      0.095   
3                     0.83                      0.119   
4                    0.941                      0.077   

  PREÇO MÍNIMO DISTRIBUIÇÃO PREÇO MÁXIMO DISTRIBUIÇÃO  \
0                    0.4201                    0.9666   
1                    0.5013                      1.05   
2                    0.5614                     1.161   
3                    0.5991                   1.22242   
4                    0.7441                    1.0317   

  COEF DE VARIAÇÃO DISTRIBUIÇÃO  
0                         0.133  
1                         0.115  
2                         0.098  
3                         0.143  
4                         0.082

## b. Estados presentes
Quais são os nomes dos estados presentes no DataFrame?  

In [25]:
df['ESTADO'].unique()

array(['DISTRITO FEDERAL', 'GOIAS', 'MATO GROSSO', 'MATO GROSSO DO SUL',
       'ALAGOAS', 'BAHIA', 'CEARA', 'MARANHAO', 'PARAIBA', 'PERNAMBUCO',
       'PIAUI', 'RIO GRANDE DO NORTE', 'SERGIPE', 'ACRE', 'AMAPA',
       'AMAZONAS', 'PARA', 'RONDONIA', 'RORAIMA', 'TOCANTINS',
       'ESPIRITO SANTO', 'MINAS GERAIS', 'RIO DE JANEIRO', 'SAO PAULO',
       'PARANA', 'RIO GRANDE DO SUL', 'SANTA CATARINA'], dtype=object)

## c. Filtragem por estado
Escolha um estado e crie um novo DataFrame apenas com os registros desse estado.  
Dê a esse DataFrame um nome apropriado, como `df_RJ` se você escolher o Rio de Janeiro.  

In [26]:
df_sp = df[df['ESTADO'] == 'SAO PAULO']

In [27]:
df_sp.head()

DATA INICIAL  DATA FINAL   REGIÃO     ESTADO           PRODUTO  \
23    2004-05-09  2004-05-15  SUDESTE  SAO PAULO  ETANOL HIDRATADO   
50    2004-05-16  2004-05-22  SUDESTE  SAO PAULO  ETANOL HIDRATADO   
77    2004-05-23  2004-05-29  SUDESTE  SAO PAULO  ETANOL HIDRATADO   
104   2004-05-30  2004-06-05  SUDESTE  SAO PAULO  ETANOL HIDRATADO   
131   2004-06-06  2004-06-12  SUDESTE  SAO PAULO  ETANOL HIDRATADO   

     NÚMERO DE POSTOS PESQUISADOS UNIDADE DE MEDIDA  PREÇO MÉDIO REVENDA  \
23                           4047              R$/l                0.768   
50                           4066              R$/l                0.766   
77                           4077              R$/l                0.823   
104                          4073              R$/l                0.887   
131                          4082              R$/l                0.894   

     DESVIO PADRÃO REVENDA  PREÇO MÍNIMO REVENDA  PREÇO MÁXIMO REVENDA  \
23                   0.085                 0.590                 1.150   
50                   0.083                 0.590                 1.150   
77                   0.071                 0.620                 1.199   
104                  0.075                 0.639                 1.359   
131                  0.071                 0.669                 1.199   

    MARGEM MÉDIA REVENDA  COEF DE VARIAÇÃO REVENDA PREÇO MÉDIO DISTRIBUIÇÃO  \
23                 0.198                     0.111                     0.57   
50                  0.19                     0.108                    0.576   
77                 0.209                     0.086                    0.614   
104                0.216                     0.085                    0.671   
131                0.205                     0.079                    0.689   

    DESVIO PADRÃO DISTRIBUIÇÃO PREÇO MÍNIMO DISTRIBUIÇÃO  \
23                       0.056                    0.4271   
50                        0.05                   0.43204   
77                       0.063                   0.34951   
104                      0.075                    0.3398   
131                      0.072                   0.47797   

    PREÇO MÁXIMO DISTRIBUIÇÃO COEF DE VARIAÇÃO DISTRIBUIÇÃO  
23                     1.4123                         0.098  
50                     0.9724                         0.087  
77                      1.412                         0.103  
104                    1.0669                         0.112  
131                    1.0669                         0.104

## d. Contagem de registros
Quantas linhas existem no DataFrame criado no item (c)?

In [28]:
df_sp.shape[0]

4819

Existem 4819 itens (linhas) no DataFrame do Estado de São Paulo.



---

## e. Conversão de datas
Converta as colunas **"DATA INICIAL"** e **"DATA FINAL"** para o tipo `datetime`.  

In [29]:
df_sp.info()

<class 'pandas.core.frame.DataFrame'>
Index: 4819 entries, 23 to 120820
Data columns (total 18 columns):
 #   Column                         Non-Null Count  Dtype  
---  ------                         --------------  -----  
 0   DATA INICIAL                   4819 non-null   object 
 1   DATA FINAL                     4819 non-null   object 
 2   REGIÃO                         4819 non-null   object 
 3   ESTADO                         4819 non-null   object 
 4   PRODUTO                        4819 non-null   object 
 5   NÚMERO DE POSTOS PESQUISADOS   4819 non-null   int64  
 6   UNIDADE DE MEDIDA              4819 non-null   object 
 7   PREÇO MÉDIO REVENDA            4819 non-null   float64
 8   DESVIO PADRÃO REVENDA          4819 non-null   float64
 9   PREÇO MÍNIMO REVENDA           4819 non-null   float64
 10  PREÇO MÁXIMO REVENDA           4819 non-null   float64
 11  MARGEM MÉDIA REVENDA           4819 non-null   object 
 12  COEF DE VARIAÇÃO REVENDA       4819 non-null   flo

In [30]:
df_sp['DATA INICIAL'] = pd.to_datetime(df_sp['DATA INICIAL'])
df_sp['DATA FINAL'] = pd.to_datetime(df_sp['DATA FINAL'])

<ipython-input-30-062c9bdfaf4b>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_sp['DATA INICIAL'] = pd.to_datetime(df_sp['DATA INICIAL'])
<ipython-input-30-062c9bdfaf4b>:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_sp['DATA FINAL'] = pd.to_datetime(df_sp['DATA FINAL'])


In [31]:
df_sp.info()

<class 'pandas.core.frame.DataFrame'>
Index: 4819 entries, 23 to 120820
Data columns (total 18 columns):
 #   Column                         Non-Null Count  Dtype         
---  ------                         --------------  -----         
 0   DATA INICIAL                   4819 non-null   datetime64[ns]
 1   DATA FINAL                     4819 non-null   datetime64[ns]
 2   REGIÃO                         4819 non-null   object        
 3   ESTADO                         4819 non-null   object        
 4   PRODUTO                        4819 non-null   object        
 5   NÚMERO DE POSTOS PESQUISADOS   4819 non-null   int64         
 6   UNIDADE DE MEDIDA              4819 non-null   object        
 7   PREÇO MÉDIO REVENDA            4819 non-null   float64       
 8   DESVIO PADRÃO REVENDA          4819 non-null   float64       
 9   PREÇO MÍNIMO REVENDA           4819 non-null   float64       
 10  PREÇO MÁXIMO REVENDA           4819 non-null   float64       
 11  MARGEM MÉDIA REVEND

## f. Conversão numérica
Converta para tipo numérico as colunas que representam preços, como:  
`PREÇO MÉDIO`, `PREÇO MÍNIMO`, `PREÇO MÁXIMO`, `PREÇO DE VENDA`, entre outras.

In [34]:
list = ['MARGEM MÉDIA REVENDA', 'PREÇO MÉDIO DISTRIBUIÇÃO', 'DESVIO PADRÃO DISTRIBUIÇÃO', 'PREÇO MÍNIMO DISTRIBUIÇÃO', 'PREÇO MÁXIMO DISTRIBUIÇÃO', 'COEF DE VARIAÇÃO DISTRIBUIÇÃO']
for atributo in list:
  df_sp[atributo] = pd.to_numeric(df_sp[atributo], errors='coerce')

<ipython-input-34-e9bdd9f71025>:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_sp[atributo] = pd.to_numeric(df_sp[atributo], errors='coerce')


In [35]:
df_sp.info()

<class 'pandas.core.frame.DataFrame'>
Index: 4819 entries, 23 to 120820
Data columns (total 18 columns):
 #   Column                         Non-Null Count  Dtype         
---  ------                         --------------  -----         
 0   DATA INICIAL                   4819 non-null   datetime64[ns]
 1   DATA FINAL                     4819 non-null   datetime64[ns]
 2   REGIÃO                         4819 non-null   object        
 3   ESTADO                         4819 non-null   object        
 4   PRODUTO                        4819 non-null   object        
 5   NÚMERO DE POSTOS PESQUISADOS   4819 non-null   int64         
 6   UNIDADE DE MEDIDA              4819 non-null   object        
 7   PREÇO MÉDIO REVENDA            4819 non-null   float64       
 8   DESVIO PADRÃO REVENDA          4819 non-null   float64       
 9   PREÇO MÍNIMO REVENDA           4819 non-null   float64       
 10  PREÇO MÁXIMO REVENDA           4819 non-null   float64       
 11  MARGEM MÉDIA REVEND

## g. Valores ausentes
Após a conversão, quantos valores ausentes (**null/NaN**) existem no DataFrame?

In [44]:
mask = df_sp['MARGEM MÉDIA REVENDA'].isnull()
df_sp[mask]

DATA INICIAL DATA FINAL   REGIÃO     ESTADO PRODUTO  \
94339    2017-11-19 2017-11-25  SUDESTE  SAO PAULO     GNV   
105018   2019-03-31 2019-04-06  SUDESTE  SAO PAULO     GNV   

        NÚMERO DE POSTOS PESQUISADOS UNIDADE DE MEDIDA  PREÇO MÉDIO REVENDA  \
94339                             20             R$/m3                2.361   
105018                            50             R$/m3                2.810   

        DESVIO PADRÃO REVENDA  PREÇO MÍNIMO REVENDA  PREÇO MÁXIMO REVENDA  \
94339                   0.140                 2.099                 2.599   
105018                  0.278                 2.479                 3.799   

        MARGEM MÉDIA REVENDA  COEF DE VARIAÇÃO REVENDA  \
94339                    NaN                     0.059   
105018                   NaN                     0.099   

        PREÇO MÉDIO DISTRIBUIÇÃO  DESVIO PADRÃO DISTRIBUIÇÃO  \
94339                        NaN                         NaN   
105018                       2.9                       0.385   

        PREÇO MÍNIMO DISTRIBUIÇÃO  PREÇO MÁXIMO DISTRIBUIÇÃO  \
94339                         NaN                        NaN   
105018                     2.2416                     3.0699   

        COEF DE VARIAÇÃO DISTRIBUIÇÃO  
94339                             NaN  
105018                          0.133

In [38]:
df_sp[mask].shape[0]

2

Existem apenas duas linhas com valores nulos

## h. Remoção de linhas com NaN
Elimine todas as linhas que contenham **valores ausentes (null/NaN)** em qualquer coluna.  

In [45]:
df_sp.dropna(inplace=True)

<ipython-input-45-661ef861f14d>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_sp.dropna(inplace=True)


In [46]:
df_sp[mask]

<ipython-input-46-4b6c9e819173>:1: UserWarning: Boolean Series key will be reindexed to match DataFrame index.
  df_sp[mask]


Empty DataFrame
Columns: [DATA INICIAL, DATA FINAL, REGIÃO, ESTADO, PRODUTO, NÚMERO DE POSTOS PESQUISADOS, UNIDADE DE MEDIDA, PREÇO MÉDIO REVENDA, DESVIO PADRÃO REVENDA, PREÇO MÍNIMO REVENDA, PREÇO MÁXIMO REVENDA, MARGEM MÉDIA REVENDA, COEF DE VARIAÇÃO REVENDA, PREÇO MÉDIO DISTRIBUIÇÃO, DESVIO PADRÃO DISTRIBUIÇÃO, PREÇO MÍNIMO DISTRIBUIÇÃO, PREÇO MÁXIMO DISTRIBUIÇÃO, COEF DE VARIAÇÃO DISTRIBUIÇÃO]
Index: []

In [47]:
df_sp.info()

<class 'pandas.core.frame.DataFrame'>
Index: 4817 entries, 23 to 120820
Data columns (total 18 columns):
 #   Column                         Non-Null Count  Dtype         
---  ------                         --------------  -----         
 0   DATA INICIAL                   4817 non-null   datetime64[ns]
 1   DATA FINAL                     4817 non-null   datetime64[ns]
 2   REGIÃO                         4817 non-null   object        
 3   ESTADO                         4817 non-null   object        
 4   PRODUTO                        4817 non-null   object        
 5   NÚMERO DE POSTOS PESQUISADOS   4817 non-null   int64         
 6   UNIDADE DE MEDIDA              4817 non-null   object        
 7   PREÇO MÉDIO REVENDA            4817 non-null   float64       
 8   DESVIO PADRÃO REVENDA          4817 non-null   float64       
 9   PREÇO MÍNIMO REVENDA           4817 non-null   float64       
 10  PREÇO MÁXIMO REVENDA           4817 non-null   float64       
 11  MARGEM MÉDIA REVEND

In [48]:
df_sp.shape[0]

4817

Como se pode ver, retiramos as duas linhas que tinham valores nulos.

## i. Exportação do resultado
Salve o DataFrame final do item (h) em um novo arquivo CSV chamado:  
`minha_pesquisa_12_abril.csv`

In [50]:
df_sp.to_csv('minha_pesquisa_12_abril.csv', index=False)

## j. Entrega
Envie o arquivo `.ipynb` (do notebook) no Moodle até o final da aula.